### PR TITLE
Fix edge cases with org join flow

### PR DESCRIPTION
**Pull request by bduffany**
:point_right: **Recommended:** [review with whitespace diff hidden](https://github.com/buildbuddy-io/buildbuddy/pull/3517/files?diff=split&w=1). (The UserDB diff is much cleaner this way) 

* Instead of `buildbuddy_server` performing multiple separate transactions to look up the group and decide whether to auto-join by domain association or submit a request to join the org, consolidate all of this logic into a single DB transaction in `RequestToJoinOrg`, and remove the `AddUserToGroup` func in UserDB.
* Fix edge case where `addUserToGroup` would not auto-promote to admin if the group was empty and there were existing membership requests which had not been approved.
* Fix edge case where we return PermissionDenied if users try to join an org after the org has claimed domain ownership and also after the user was created (this was because of the admin auth check in AddUserToGroup - this PR removes that func entirely).

---

**Version bump**: Patch

### CODE REVIEW

File: enterprise/server/backends/userdb/userdb.go [hunk 1]
```diff
@@ -639,15 +643,26 @@ func (d *UserDB) DeleteGroupGitHubToken(ctx context.Context, groupID string) err
...
+		return d.addUserToGroup(tx, userID, groupID)
 	})
 }
 
-func (d *UserDB) addUserToGroup(ctx context.Context, tx *db.DB, userID, groupID string) error {
+func (d *UserDB) addUserToGroup(tx *db.DB, userID, groupID string) error {
+	row := &struct{ Count int64 }{}
+	err := tx.Raw(`
+		SELECT COUNT(*) AS count FROM UserGroups
+		WHERE group_group_id = ? AND membership_status = ?
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:655]: should this query also look at userID?

File: enterprise/server/backends/userdb/userdb.go [hunk 2]
```diff
@@ -656,60 +671,73 @@ func (d *UserDB) addUserToGroup(ctx context.Context, tx *db.DB, userID, groupID
...
+		}
+
+		// Check if there's an existing request and return AlreadyExists if so.
+		existing, err := getUserGroup(tx, userID, groupID)
+		if err != nil {
+			return err
+		}
+		if existing != nil {
 			return status.AlreadyExistsError("You've already requested to join this organization.")
 		}
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:729]: do we want to distinguish between already requested and already a member?

File: enterprise/server/backends/userdb/userdb.go [hunk 3]
```diff
@@ -656,60 +671,73 @@ func (d *UserDB) addUserToGroup(ctx context.Context, tx *db.DB, userID, groupID
...
-		return status.InvalidArgumentError("User ID is required.")
+		return 0, status.InvalidArgumentError("User ID is required.")
 	}
 	if groupID == "" {
-		return status.InvalidArgumentError("Group ID is required.")
+		return 0, status.InvalidArgumentError("Group ID is required.")
 	}
-	return d.h.Transaction(ctx, func(tx *db.DB) error {
-		existing, err := getUserGroup(tx, userID, groupID)
+	var newStatus grpb.GroupMembershipStatus
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:704]: might be clearer to call this membershipStatus

File: enterprise/server/backends/userdb/userdb.go [hunk 4]
```diff
@@ -1162,3 +1190,11 @@ func hasAdminOnlyCapabilities(capabilities []akpb.ApiKey_Capability) bool {
 	}
 	return false
 }
+
+func getEmailDomain(email string) string {
+	chunks := strings.Split(email, "@")
+	if len(chunks) == 0 {
+		return ""
+	}
+	return chunks[len(chunks)-1]
```
Comment (tylerwilliams) [inline enterprise/server/backends/userdb/userdb.go:1199]: No strong preference here, but another way to do this if you expect two parts is to use `strings.SplitN`

```
parts := strings.SplitN(email, "@", 2)
if len(parts) != 2 {
  return ""
}
return parts[1]
```

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/backends/userdb/userdb.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/backends/userdb/userdb.go:731]: Woops, missed this comment. We do distinguish these 2 cases - the "already a member" case is just moved into the `addUserToGroup` func

Comment (bduffany) [reply enterprise/server/backends/userdb/userdb.go:731]: Actually nvm - that addUserToGroup is only called in the owned-domain case. Will fix and add a test

Comment (bduffany) [reply enterprise/server/backends/userdb/userdb.go]: Done - PTAL

**Review by tylerwilliams — APPROVED** [review]

**Pull request merged**